llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -1471,10 +1470,11 @@ std::string NVPTXTargetLowering::getPrototype(
       continue;
     }
 
+    // Indirect calls need strict ABI alignment so we disable optimizations.
     Type *ETy = Args[i].IndirectType;
     Align InitialAlign = Outs[OIdx].Flags.getNonZeroByValAlign();
     Align ParamByValAlign =
-        getFunctionByValParamAlign(F, ETy, InitialAlign, DL);
+        getFunctionByValParamAlign(/*F=*/nullptr, ETy, InitialAlign, DL);
```

Comment:
The comment above needs to mantion that it's the `nullptr` here that forces strict ABI-compliant alignment. Otherwise it's not clear which part of the code controls this.

Ideally, it would be even nicer to make "ABI-compliant alignment" a parameter.

Reply:
This doesn't really make sense to me as a fix. Indirect calls shouldn't require special handling vs. direct

Reply:
The NVPTX backend thinks that `internal` functions can have their ABI modified, so they up-scale alignment on things. This doesn't work for these indirect calls, which seem to strictly require the ABI.